servicetalk-grpc-netty/src/main/java/io/servicetalk/grpc/netty/DefaultGrpcClientBuilder.java

```diff
@@ -168,4 +227,91 @@ public void doAppendHttpClientFilter(final Predicate<StreamingHttpRequest> predi
                                          final StreamingHttpClientFilterFactory factory) {
         httpClientBuilder.appendClientFilter(predicate, factory);
     }
+
+    private TimeoutFromRequest grpcTimeout() {
+
+        return new TimeoutFromRequest() {
+            /**
+             * Return the timeout duration extracted from the GRPC timeout HTTP header if present, from the async
+             * context if present or the default timeout.
+             *
+             * @param request The HTTP request to be used as source of the GRPC timeout header
+             * @return The non-negative timeout duration which may be null
+             */
+            @Nullable
+            public Duration apply(HttpRequestMetaData request) {
+                Duration duration = readTimeoutHeader(request);
+
+                if (null != duration) {
+                    // use timeout from request
+                    return duration;
+                }
+
+                Instant deadline = AsyncContext.get(GRPC_DEADLINE_KEY);
+
+                if (null != deadline) {
+                    // Use remaining time until deadline from async context
+                    return Duration.between(Instant.now(), deadline);
+                }
+
+                return defaultTimeout;
+            }
+
+            @Override
+            public HttpExecutionStrategy influenceStrategy(final HttpExecutionStrategy strategy) {
+                return strategy;
+            }
+        };
+    }
+
+    /**
+     * Extract the timeout duration from the HTTP headers if present.
+     *
+     * @param request The HTTP request to be used as source of the GRPC timeout header
+     * @return The non-negative timeout duration which may null if not present or malformed
+     */
+    @Nullable
+    static Duration readTimeoutHeader(HttpRequestMetaData request) {
+        return parseTimeoutHeader(request.headers().get(GRPC_TIMEOUT_HEADER_KEY));
+    }
+
+    /**
+     * Parse a gRPC HTTP timeout header value as a duration if present.
+     *
+     * @param value the text value of {@link #GRPC_TIMEOUT_HEADER_KEY} header to be parsed to a timeout duration or
+     * null if header was present
+     * @return The non-negative timeout duration which may null if not present or malformed
+     */
+    static @Nullable Duration parseTimeoutHeader(@Nullable CharSequence value) {
+        if (null != value && value.length() >= 2 && value.length() <= 9) {
```

Comment:
We should probably throw if the value is malformed instead of returning a `null`

Reply:
Yes, reasonable. Is it acceptable to throw IllegalArgumentException in this and the other malformed header cases?

Reply:
In this case the IllegalArgumentException would be thrown from the `Single.defer()` returned by `TimeoutHttpRequesterFilter.request()` so will show upon the subscribe.